File: runtimes/swiftui/Sources/StyleConverterRuntime/StyleEngine/scrolling/LineClampCap.swift
//
//  LineClampCap.swift
//  StyleEngine/scrolling — Wave 46 (lane Y1): the BLOCK-LEVEL line-clamp
//  height cap, iOS twin of Compose's scrolling/LineClampCap.kt (wave 41).
//
//  css-overflow-4 §5 expands `line-clamp: <n>` into `max-lines: <n>` +
//  `block-ellipsis: auto` + `continue: discard`: the clamp container keeps
//  its first <n> LINE BOXES and discards everything after them, and §4.3
//  makes the discarded content unpaintable ("is not rendered"). The iOS
//  leaf path honors this for a component whose text lives in ONE label
//  (PlaceholderLabel's inner `.lineLimit(textConfig.lineClampLimit)`), but
//  a clamp root whose line boxes live in CHILD components — `<span>` runs,
//  nested `<p>`/`<div>` blocks, an inline-block atom — rendered every child
//  unclamped: the wave-45 iOS captures show block-ellipsis-012's 1-line
//  box 3 lines tall (iOS-ref 0.9245), -022's 2-line box 3 lines tall
//  (0.9451), -027/-031/-015 5 lines where the ref closes after 2
//  (0.868/0.870/0.870) — all tests whose Android twin, capped since wave
//  41, passes. SwiftUI's `.lineLimit` is an ENVIRONMENT value: each child
//  Text clamps itself at N, the container never does. So the container
//  must cap: this file is the pure half — the wire gate, the line-box
//  pick, the cap height and the cap decision — pinned by
//  LineClampCapTests without a render surface; LineClampHeightCapLayout
//  is the layout half; LineClampCensus walks the children for the
//  non-uniform case.
//
//  Chain slot (the renderer seam): the cap wraps the CONTENT box — inside
//  the border inset / padding / sizing chain (StyleBuilder.applyBoxDecoration),
//  i.e. on `flowContainer(style:)` — so the budget is N line boxes only;
//  the Compose twin sits at its step-7 overflow node and therefore adds
//  the padding + border band that nest inside that node. Same model,
//  different measuring point.
//

import CoreGraphics
import Foundation

enum LineClampCap {

    // Tolerance (pt) between the estimated cap and a measurement that is
    // ALREADY clamped by the leaf label. A leaf's N clamped lines measure
    // N × the same pinned line box this file computes, so the two agree up
    // to rounding of fractional line heights (monospace 13px × 1.25 =
    // 16.25pt/line); 2pt absorbs that rounding so the cap NEVER re-clips a
    // correctly clamped leaf (block-ellipsis-013's passing 2-line box stays
    // byte-identical), while a genuine overflow is always ≥ one extra line
    // box (≥ 16pt) past the cap and still trips it. Same constant as
    // Compose's LineClampCap.SLACK_PX — the cross-native pin.
    static let slackPx: CGFloat = 2

    /// The clamp line count, gated on the RAW wire shape.
    ///
    /// Only the `{"type":"lines","count":N}` variant caps (the shorthand's
    /// `<integer [1,∞]>` grammar, css-overflow-4 §5.1 — the converter's
    /// LineClampPropertyParser emits exactly this for it). `none` (clamp
    /// off), `auto` (clamp at the box's own block-size constraint — no
    /// fixed line count to turn into a height) and the legacy bare-integer
    /// shape deliberately return nil so this cap never guesses; the read
    /// is delegated to LineClampExtractor so the leaf's `.lineLimit` and
    /// this cap can never disagree about the count.
    static func linesCount(_ properties: [IRProperty]) -> Int? {
        // The LAST LineClamp declaration wins (cascade order) — the
        // extractor folds the whole list in order, exactly like that.
        guard let last = properties.last(where: { $0.type == LineClampProperty.name }),
              // Gate on the sealed `lines` discriminator: the extractor
              // ALSO accepts the legacy bare integer, which this cap
              // refuses (Compose parity — only the live wire shape caps).
              case .object(let o) = last.data,
              o["type"]?.stringValue == "lines"
        else { return nil }
        // The ≥1 guard lives in the extractor (css-overflow-4 §5.1).
        return LineClampExtractor.extract(from: properties)?.lines
    }

    /// True when the declaration forbids a marker (`no-ellipsis` / `""`,
    /// LineClampConfig.markerSuppressed). SwiftUI's `.lineLimit` tail
    /// truncation ALWAYS paints "…", so a marker-less clamp cannot ride the
    /// leaf path: the renderer routes it through the cap instead — the text
    /// lays out every line and the cap discards the tail without a glyph
    /// (block-ellipsis-023's iOS capture paints the forbidden "…").
    static func markerSuppressed(_ properties: [IRProperty]) -> Bool {
        LineClampExtractor.extract(from: properties)?.markerSuppressed ?? false
    }

    /// The leaf label's `.lineLimit` argument: the clamp count as today,
    /// or nil when the marker is suppressed — SwiftUI cannot truncate
    /// without painting "…", so the marker-less clamp is enforced by the
    /// cap/clip route alone (the leaf must lay out ALL its lines for the
    /// first N to keep their natural geometry, css-overflow-4 §4.3). Pure,
    /// so the seam is a one-call fold on TextConfig.lineClampLimit.
    static func leafLineLimit(_ limit: Int?, properties: [IRProperty]) -> Int? {
        markerSuppressed(properties) ? nil : limit
    }

    /// One line box in pt: a DECLARED line-height wins verbatim (css-inline-3
    /// §4.2 — line-clamp-005 pins `font: 16px/32px` → 32pt boxes); otherwise
    /// the corpus-v4.1 composed default — font-size × 1.25, the REF_LINE_
    /// HEIGHT ratio shared by all four surfaces (ComponentRenderer.
    /// wptRefLineHeightRatio; Compose composedDefaultLineHeightPx with
    /// composedWpt = true — the twin ALSO pins the ratio regardless of
    /// capture mode, because a cap needs a number and the face's natural
    /// metrics are not known without a layout pass). A non-positive font
    /// size can only come from a broken wire — refuse to build a 0-height
    /// cap out of it (nil).
    static func lineBoxPx(declaredLineHeightPx: CGFloat?, fontSizePx: CGFloat) -> CGFloat? {
        // Declared px line-height → the line box, verbatim.
        if let lh = declaredLineHeightPx, lh > 0 { return lh }
        // Calibrated default: the ref's unitless 1.25 × THIS run's size.
        guard fontSizePx > 0 else { return nil }
        return fontSizePx * ComponentRenderer.wptRefLineHeightRatio
    }

    /// The uniform cap: N line boxes of one height — the Compose model,
    /// and the fallback when LineClampCensus cannot prove the children's
    /// line boxes (wrapping runs of unknown count).
    static func capPx(lines: Int, lineBoxPx: CGFloat) -> CGFloat? {
        // The grammar is <integer [1,∞]>; a 0-line cap would blank the box.
        guard lines >= 1, lineBoxPx > 0 else { return nil }
        return CGFloat(lines) * lineBoxPx
    }

    /// The pure cap decision (pinnable without a layout pass): the
    /// measured height survives untouched unless it exceeds the cap by
    /// more than [slackPx] — i.e. the content genuinely laid out extra
    /// line boxes the clamp must discard. The cap itself is ceil'd so a
    /// fractional line box (16.25pt monospace lines) never loses its last
    /// physical pixel row — byte-parallel to Compose's cappedHeight.
    static func cappedHeight(measuredPx: CGFloat, capPx: CGFloat) -> CGFloat {
        // The integer cap the box reports when clamping fires.
        let cap = ceil(capPx)
        // Within tolerance → the leaf path (or short content) already
        // satisfies the clamp: keep the measurement byte-identically.
        return measuredPx > cap + slackPx ? cap : measuredPx
    }
}

File: runtimes/compose/src/main/java/com/styleconverter/runtime/scrolling/LineClampCap.kt
package com.styleconverter.runtime.scrolling

// Wave 41 (lane T6) — the BLOCK-LEVEL line-clamp height cap.
//
// css-overflow-4 §5 expands `line-clamp: <n>` into `max-lines: <n>` +
// `block-ellipsis: auto` + `continue: discard`: the clamp container keeps
// its first <n> LINE BOXES and discards everything after them, and §4.3
// makes the discarded content unpaintable (it "is not rendered"). The
// leaf path already honors this for a component whose text lives in ONE
// Compose Text (ComponentRenderer.placeholderMaxLines → Text(maxLines)),
// but a clamp root whose line boxes live in CHILD components — `<span>`
// runs, nested `<p>`/`<div>` blocks — rendered every child unclamped:
// wave-40 pixels show block-ellipsis-012's 1-line box rendered 3 lines
// tall on Android (border rows 16..66 vs the ref's 16..33), and
// block-ellipsis-032's three 1-line boxes rendered 4 stacked runs each
// (Android SSIM 0.741). Text maxLines cannot reach children, so the
// container itself must cap: this file computes the used cap height
// (N line boxes + the vertical bands that live inside the overflow node)
// and provides the measure-then-cap layout modifier OverflowApplier
// chains whenever a fixed-count LineClamp property is on the wire.

import androidx.compose.ui.Modifier
import androidx.compose.ui.layout.layout
import androidx.compose.ui.unit.Constraints
import com.styleconverter.runtime.StyleApplier
import com.styleconverter.runtime.borders.sides.BorderSideExtractor
import com.styleconverter.runtime.core.renderer.composedDefaultLineHeightPx
import com.styleconverter.runtime.spacing.SpacingContext
import com.styleconverter.runtime.spacing.SpacingExtractor
import com.styleconverter.runtime.spacing.resolveToDp
import com.styleconverter.runtime.typography.LineClampWire
import com.styleconverter.runtime.typography.TypographyExtractor
import kotlin.math.ceil
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.floatOrNull

object LineClampCap {

    // Tolerance (px) between the estimated cap and a measurement that is
    // ALREADY clamped by the leaf Text path. A leaf's N clamped lines
    // measure N × the same pinned line box this file computes, so the two
    // agree up to Int-rounding of fractional line heights (monospace
    // 13px × 1.25 = 16.25px/line); 2px absorbs that rounding so the cap
    // NEVER re-clips a correctly clamped leaf (block-ellipsis-013 keeps
    // its passing 34px Android box byte-identically), while a genuine
    // overflow is always ≥ one extra line box (≥ 16px) past the cap and
    // still trips it.
    const val SLACK_PX: Float = 2f

    /**
     * The clamp line count, gated on the RAW wire shape.
     *
     * Only the `{"type":"lines","count":N}` variant caps (the shorthand's
     * `<integer [1,∞]>` grammar, css-overflow-4 §5.1 — the converter's
     * LineClampPropertyParser emits exactly this for it). `none` (clamp
     * off), `auto` (clamp at the box's own block-size constraint — no
     * fixed line count to turn into a height) and any legacy shape
     * deliberately return null so this cap never guesses. The count read
     * mirrors TypographyExtractor.extractLineClamp (the typography lane's
     * wire read) including its ≥1 guard — sub-1 counts are outside the
     * grammar and mean "clamp off", never "0-height box".
     */
    fun linesCount(properties: List<Pair<String, JsonElement?>>): Int? {
        // The LAST LineClamp declaration wins, mirroring CSS cascade order
        // exactly like the typography extractor's fold over the same list.
        val data = properties.lastOrNull { it.first == "LineClamp" }?.second
            ?: return null // no declaration → no cap; the common fast path.
        // Non-object payloads are legacy shapes the cap refuses (see doc).
        val obj = data as? JsonObject ?: return null
        // The sealed-variant discriminator the converter serializes.
        val type = (obj["type"] as? JsonPrimitive)?.contentOrNull
        // Gate: only the fixed-count variant maps to a fixed cap height.
        if (type != "lines") return null
        // IRNumber serializes the count as a float ("count": 2.0); the
        // ≥1 guard is the css-overflow-4 §5.1 <integer [1,∞]> grammar.
        return (obj["count"] as? JsonPrimitive)?.floatOrNull?.toInt()
            ?.takeIf { it >= 1 }
    }

    /**
     * Retrospective R3 (A5#4) — true when the declaration forbids a marker
     * (`line-clamp: 4 no-ellipsis` / `4 ""`, css-overflow-4 §4.2). Twin of
     * Swift's LineClampCap.markerSuppressed, delegated to the ONE Compose
     * wire reader ([LineClampWire]) so the cap, the leaf Text and the
     * typography config can never disagree about the marker. Unlike Swift,
     * the Compose cap needs no re-route on it: Text(maxLines, Clip) already
     * discards without a glyph, so this is exposed for the renderer's
     * overflow DECISION (Clip, never Ellipsis) and for the JVM pins.
     */
    fun markerSuppressed(properties: List<Pair<String, JsonElement?>>): Boolean =
        LineClampWire.markerSuppressedInPairs(properties)

    /**
     * The full cap height in px for this component, or null when no cap
     * applies. The cap is what the STEP-7 overflow node may measure:
     *
     *   N × lineBox  +  vertical CSS padding  +  vertical border band
     *
     * because the modifier chain (StyleApplier.applyConfig steps 4-8 +
     * ComponentRenderer's inner chain) nests the css padding (step 8),
     * the placeholder floor and the border-band content inset all INSIDE
     * the step-7 overflow position — the wave-40 leaf calibration pins
     * it: block-ellipsis-013's PASSING Android box measures 35px against a
     * 34.5px cap (2 lines × 16.25 — monospace 13px × the 1.25 ref pin — +
     * 2 × 1px border band = 34.5, ceil → 35; the wave49-final capture and
     * the frozen ref both ink rows 16..50, and LineClampCapTest's
     * "measurement within slack" pin carries the same 35-vs-34.5 numbers;
     * retro A3#13 corrected the "34px" arithmetic that stood here).
     */
    fun capPx(properties: List<Pair<String, JsonElement?>>): Float? {
        // Fixed-count clamp on the wire, else nothing to do (fast bail —
        // this runs inside every component's overflow extraction).
        val lines = linesCount(properties) ?: return null
        // Reuse the typography lane's OWN resolution of font-size and
        // line-height (the same public function StyleApplier already runs
        // for this very component), so cap metrics and leaf-Text metrics
        // can never disagree: fontSize carries the monospace-13px UA
        // quirk (MonospaceUAFontSize) and lineHeight the multiplier ×
        // font-size resolution — both in sp, and the runtime's
        // px==sp==dp space makes .value the px count.
        val typo = TypographyExtractor.extractTypographyConfig(properties)
        // The element's used font size in px: declared (or UA-monospace
        // 13px, folded in by the extractor) else the CSS body default 16.
        val fontPx = typo.fontSize?.takeIf { it.isSp }?.value ?: 16f
        // One line box: a DECLARED line-height wins verbatim (css-inline-3
        // §4.2 — line-clamp-005 pins `font: 16px/32px` → 32px boxes);
        // otherwise the corpus-v4.1 composed default — font-size × 1.25,
        // the REF_LINE_HEIGHT pin shared by all four surfaces (see
        // core/renderer WptCaptureMode.REF_DEFAULT_FONT_LINE_HEIGHT_RATIO).
        val lineBox = typo.lineHeight?.takeIf { it.isSp }?.value
            ?: composedDefaultLineHeightPx(composedWpt = true, fontSizePx = fontPx)
        // A non-positive line box can only come from a broken wire —
        // refuse to build a 0-height cap out of it.
        if (lineBox <= 0f) return null
        // Vertical CSS padding — resolved EXACTLY like the padding
        // modifier resolves it (same extractor + resolver + the default
        // context StyleApplier.placeholderFloorMinSize uses), so the band
        // the cap budgets is the band the chain actually lays out.
        val resolved = SpacingExtractor.extractPaddingConfig(properties)
            .resolve(isRtl = false)
        // The default resolution context: px-exact values dominate the
        // corpus; relative units resolve against the same defaults the
        // floor math uses, so cap and floor never disagree.
        val ctx = SpacingContext()
        // Negative padding is not a thing in CSS — clamp at 0 exactly
        // like the floor math does.
        val padTop = resolveToDp(resolved.top, ctx).value.coerceAtLeast(0f)
        val padBottom = resolveToDp(resolved.bottom, ctx).value.coerceAtLeast(0f)
        // Vertical border band — the content inset borderContentInset
        // chains inside the overflow node; borderBandInsets is the SAME
        // per-side math (a side reserves width only for a VISIBLE
        // border), returned in (start, top, end, bottom) order.
        val band = StyleApplier.borderBandInsets(
            BorderSideExtractor.extractBorderConfig(properties)
        )
        // Sum the block-axis pieces: N line boxes + padding + border band.
        return lines * lineBox + padTop + padBottom + band[1].value + band[3].value
    }

    /**
     * The pure cap decision (JVM-pinnable without a draw pass): the
     * measured height survives untouched unless it exceeds the cap by
     * more than [SLACK_PX] — i.e. the content genuinely laid out extra
     * line boxes the clamp must discard. The cap itself is ceil'd so a
     * fractional line box (16.25px monospace lines) never loses its last
     * physical pixel row to Int truncation.
     */
    fun cappedHeight(measuredPx: Int, capPx: Float): Int {
        // The integer cap the box reports when clamping fires.
        val cap = ceil(capPx).toInt()
        // Within tolerance → the leaf path (or short content) already
        // satisfies the clamp: keep the measurement byte-identically.
        return if (measuredPx > cap + SLACK_PX) cap else measuredPx
    }
}

/**
 * Cap this node's LAYOUT height at [capPx] (the N-line-box budget), while
 * measuring content on an UNBOUNDED block axis so text keeps its natural
 * line layout — the visible prefix must be the same first N lines the
 * unclamped layout produces (css-overflow-4 §5.3.2: discarding content must
 * not re-flow what precedes it). Chain an [axisSelectiveClip] Y-clip
 * OUTSIDE this node (OverflowApplier does) so the discarded lines' ink is
 * actually unpaintable, and never clip X: a clamp does not create an
 * inline-axis clip (block-ellipsis-013's preserved first line paints past
 * the border box's right edge in the ref, and must keep doing so).
 */
fun Modifier.lineClampHeightCap(capPx: Float): Modifier =
    // Modifier.layout: a measure/report-size wrapper — no layer, no draw.
    layout { measurable, constraints ->
        // Unbound the block axis (keep the inline axis exactly as given so
        // wrapping width is untouched): content lays out ALL its lines.
        val inner = constraints.copy(minHeight = 0, maxHeight = Constraints.Infinity)
        // One measurement — the natural (unclamped) content geometry.
        val placeable = measurable.measure(inner)
        // Pure decision: keep the measurement, or discard past the cap.
        val capped = LineClampCap.cappedHeight(placeable.height, capPx)
        // Whatever was decided must still honor the INCOMING constraints —
        // an explicit `height:` (applied outer, chain step 4) wins over
        // the cap exactly like CSS used-height wins over max-lines.
        val h = capped.coerceIn(constraints.minHeight, constraints.maxHeight)
        // Report the capped box; place content at the origin so its first
        // N lines fill it and the discarded tail hangs below (the outer
        // Y-clip erases that ink, css-overflow-4 §5.3.2).
        layout(placeable.width, h) { placeable.place(0, 0) }
    }
